src/Argument.h

```diff
@@ -22,19 +23,63 @@ struct Argument {
     std::string name;
 
     /** An argument is either a primitive type (for parameters), or a
-     * buffer pointer. If 'is_buffer' is true, then 'type' should be
-     * ignored.
+     * buffer pointer.
+     *
+     * If kind == Scalar, then type fully encodes the expected type
+     * of the scalar argument.
+     *
+     * If kind == Buffer, then type.bytes() should be used to determine
+     * elem_size of the buffer; additionally, type.code *should* reflect
+     * the expected interpretation of the buffer data (e.g. float vs int),
+     * but there is no runtime enforcement of this at present.
      */
-    bool is_buffer;
+    enum Kind {
+        Scalar,
+        Buffer
+    };
+    Kind kind;
 
-    /** If this is a scalar parameter, then this is its type */
+    /** If kind == Buffer is true, this is the dimensionality of the buffer.
+     * If kind == Scalar is false, this value is ignored (and should always be set to zero) */
+    uint8_t dimensions;
+
+    /** If this is a scalar parameter, then this is its type.
+     *
+     * If this is a buffer parameter, this is used to determine elem_size
+     * of the buffer_t.
+     *
+     * Note that type.width should always be 1 here. */
     Type type;
 
-    Argument() : is_buffer(false) {}
-    Argument(const std::string &_name, bool _is_buffer, Type _type) :
-        name(_name), is_buffer(_is_buffer), type(_type) {
+    /** If this is a scalar parameter, then these are its default, min, max values.
+     * By default, they are left unset, implying "no default, no min, no max".
+     *
+     * TODO: Expr is constrained to a 32-bit value (either int or float),
+     * so this constrains the default for int64 and double. (The same constraint
```

Comment:
They need not be constants, so an int64 could conceivably be encoded as (cast<int64_t>(high_bytes) << 32) | low_bytes